hdfs/util.py

```diff
@@ -136,6 +137,105 @@ def write(self, chunk):
       self._queue.put(chunk)
 
 
+class BoundedAsyncWriter(AsyncWriter):
+ 
+  """A Bounded asynchronous publisher-consumer.
+ 
+  :param consumer:    Function which takes a single generator as argument.
+  :param buffer_size: Number of entities that are buffered. When this number is exeeded,
+                      write will block untill some of the entities are consumed
+ 
+  This class extends AsyncWriter with a fixed buffer size. If the buffer size is exeeded, 
+  writes will be blocked untill some of the buffer is consumed:
+ 
+  """
+ 
+  # Expected by pandas to write csv files (https://github.com/mtth/hdfs/pull/130).
+  __iter__ = None
+ 
+  def __init__(self, consumer, buffer_size=1024):
+    super().__init__(consumer)
+    self._content_length = 0
+    self._content_max = buffer_size
+    self._content_lock = Lock()
+
+  @property
+  def is_full(self):
+    return self._content_lock.locked()
+    
+  def __enter__(self):
+ 
+    if self._queue:
+      raise ValueError('Cannot nest contexts.')
+    
+    self._queue = Queue()
+    self._err = None
+ 
+    self._content_length = 0
+ 
+    def reader(queue):
+      """Generator read by the consumer."""
+      while True:
+        chunk = queue.get()
+        if chunk is None:
+          break
+        
+        self._content_length -= len(chunk)
+        if self._content_lock.locked() and self._content_length < self._content_max:
+          _logger.debug("releasing lock from reader")
+          _logger.debug(f"Current buffer size: {self._content_length}")
+          try:
+            self._content_lock.release()
+          except RuntimeError:
+            pass
```

Comment:
This is a big code smell. Is there a way to avoid this? Locks are best used via `with`, perhaps it can help us here.